src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs

```diff
@@ -0,0 +1,164 @@
+        }
+
+        internal static JsonConverter GetBuiltinConverter(Type typeToConvert)
+        {
+            if (s_defaultSimpleConverters == null || s_defaultFactoryConverters == null)
+            {
+                // (De)serialization using serializer's options-based methods has not yet occurred, so the built-in converters are not rooted.
+                // Even though source-gen code paths do not call this method <i.e. JsonSerializerOptions.GetConverter(Type)>, we do not root all the
+                // built-in converters here since we fetch converters for any type included for source generation from the binded context (Priority 1).
+                Debug.Assert(s_defaultSimpleConverters == null);
+                Debug.Assert(s_defaultFactoryConverters == null);
+                ThrowHelper.ThrowNotSupportedException_BuiltInConvertersNotRooted(typeToConvert);
+                return null!;
+            }
+
+            if (s_defaultSimpleConverters.TryGetValue(typeToConvert, out JsonConverter? foundConverter))
+            {
+                return foundConverter;
+            }
+            else
+            {
+                foreach (JsonConverter item in s_defaultFactoryConverters)
+                {
+                    if (item.CanConvert(typeToConvert))
+                    {
+                        return item;
+                    }
+                }
+
+                Debug.Fail("Since the object and IEnumerable converters cover all types, we should have a converter.");
+                return null!;
+            }
+        }
+
+        internal static JsonConverter? GetSimpleBuiltinConverter(Type typeToConvert)
+        {
+            JsonConverter? converter = null;
+            s_defaultSimpleConverters?.TryGetValue(typeToConvert, out converter);
+            return converter;
+        }
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        private static void RootReflectionSerializerDependencies()
+        {
+            // s_defaultFactoryConverters is the last field assigned.
+            // Use it as the sentinel to ensure that all dependencies are initialized.
+            if (s_defaultFactoryConverters == null)
+            {
+                s_defaultSimpleConverters = GetDefaultSimpleConverters();
+                s_defaultFactoryConverters = GetDefaultFactoryConverters();
+            }
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "The ctor is marked with RequiresUnreferencedCode.")]
+        internal JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
+        {
+            JsonTypeInfo.ValidateType(type, null, null, options);
+
+            MethodInfo methodInfo = typeof(JsonSerializerOptions).GetMethod(nameof(JsonSerializerOptions.CreateReflectionJsonTypeInfo), BindingFlags.NonPublic | BindingFlags.Instance)!;
```

Comment:
Why isn't `JsonSerializerOptions.CreateReflectionJsonTypeInfo` moved to this class as well?